pep-9999.txt

```diff
@@ -0,0 +1,171 @@
+        Donald Stufft <donald@stufft.io>, Nick Coghlan <ncoghlan@gmail.com>
+Status: Draft
+Type: Standards Track
+Content-Type: text/x-rst
+Created: 16-May-2018
+Python-Version: 3.8
+
+
+
+Abstract
+========
+
+This PEP proposes to add to Python a mechanism to have a ``__pypackages__``
+directory  per project to have its own independent set of installed packages.
+This will share the standard library with the base installed Python. This will
+avoid the steps to create or activate or deactivate "virtual environments". Any
+Python scripts will use the ``__pypackages__`` from the base directory of the 
+script.
+
+
+
+Motivation
+==========
+
+Python virtual environments have become an essential part of development and
+teaching workflow in the community, but at the same time, they create a
+barrier to entry for many folks. The following are few examples of the issues
+people get into while being introduced to Python (or programming for the first
+time).
+
+- How virtual environments work is a lot of information for anyone new. It takes
+  a lot of extra time and effort to explain them.
+
+- Different platforms and shell environments require different sets of commands
+  to work on the virtual environments. Any workshop or teaching environment with
+  people coming with different operating systems installed on the laptops create a
+  lot of confusion among the participants.
+
+- Virtual environments need to be activated on each opened terminal, if someone
+  creates/opens a new terminal, that by default does not get the same environment
+  as in a previous terminal with virtual environment activated.
+
+
+Specification
+=============
+
+When the Python binary is executed, it attempts to determine its
+prefix (which it stores in ``sys.prefix``), which is then used to find
+the standard library and other key files, and by the ``site`` module
+to determine the location of the site-package directories.  Currently
+the prefix is found (assuming ``PYTHONHOME`` is not set) by first
+walking up the filesystem tree looking for a marker file (``os.py``)
+that signifies the presence of the standard library, and if none is
+found, falling back to the build-time prefix hard coded in the binary.
+
+This PEP proposes to add a new step in this process. If a ``__pypackages__``
+directory is found in the current working directory, then it will be included in
+``sys.path`` after the current working directory and just before the system
+site-packages. This way, if the Python executable starts in the given project
+directory, it will automatically find all the dependencies inside of
+``__pypackages__``.
+
+In case Python scripts, Python will try to find ``__pypackages__`` in the
+base directory of the script. If found, then it will be used, otherwise the
+Python will behave as it does currently.
+
+If ``pip`` finds the ``__pypackages__`` directory in the current directory,
+it will, by default, install any packages there. There will also be a command
```

Comment:
Should we specify other file types here? Bin files, etc? It appears the intent is something like:

```
    foo
        __pypackages__
            Lib
                site-packages
                    bottle
            bin
                a-script-that-comes-with-bottle
        myscript.py
```

Which means that people can execute that script as ``__pypackages__/bin/a-script-that-comes-with-bottle``, and then pip would likely introduce a ``run`` sub command like ``pip run a-script-that-comes-with-bottle`` that does the location for the user.

Reply:
IMO, recommending `python -m` is better than adding `pip run`.
(Even better if `py -m` becomes possible everywhere)

Reply:
``-m`` and something like ``pip run`` do different things. ``-m`` is completely useless if the thing you're trying to execute is a script instead of a module (and the script support in Python's packaging supports things like binaries, bash scripts, etc).

Projects are going to ship scripts still (and there are a large corpus of existing ones that already do), so having an answer of "where do we put these script files" is important.

Reply:
Thinking about this some more, it seems like the ``Lib/site-packages`` folder is sort of redundant, since I don't think anything else would ever go into ``Lib/`` besides ``site-packages``,  and every thing else would just be under ``__pypackages__``. Maybe something like (ignoring the question about ABIs below):

```
__pypackages__
├── bin
├── data
├── include
└── site-packages
```

There are [8 different categories of paths in sysconfig](https://docs.python.org/3/library/sysconfig.html#installation-paths):

* ``stdlib`` - This isn't important for the ``__pypackages__`` directory, because you can't put the standard library in there anyways.
* ``platstdlib`` - Same as ``stdlib``.
* ``purelib`` - This is the primary purpose of this folder, so it obviously makes sense to keep it, and this maps to ``site-packages`` above.
* ``platlib`` - This doesn't really make sense for this, and it's pretty much just used on some platform provided Pythons. I think it would just be ``site-packages`` as well.
* ``include`` - We need some place to put header files shipped in packages, this corresponds to ``include`` above.
* ``platinclude`` - Much like ``platlib``, I don't think this is super useful and would just put them in ``include``.
* ``scripts`` - A place to put scripts, this corresponds to ``bin`` above.
* ``data`` - A place to put data files, this corresponds to ``data`` above.

So I think we can easily boil this down to just 4 standard directories, and map the 8 categories as specified.